### PR TITLE
Add test plan for oneapi_non_uniform_groups extension

### PR DESCRIPTION
**Pull request by steffenlarsen**
This commit adds a test plan for the [oneapi_non_uniform_groups](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_non_uniform_groups.asciidoc) oneAPI extension.

### CODE REVIEW

File: test_plans/non_uniform_groups.asciidoc
Comment (keryell) [inline test_plans/non_uniform_groups.asciidoc:7]: URL not working.

Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc]: My bad! It should work now.

**Comment by gmlueck** [comment]
@Pennycook could I delegate my review role to you for this test plan?  I think you'd have better insight since you were the original author of this extension spec.  (I tried adding you in the "Reviewers" section, but for some reason GitHub doesn't find you there.)

**Comment by Pennycook** [comment]
> @Pennycook could I delegate my review role to you for this test plan? I think you'd have better insight since you were the original author of this extension spec. (I tried adding you in the "Reviewers" section, but for some reason GitHub doesn't find you there.)

I think I can't be officially added as a reviewer because I'm not associated with the SYCL-CTS repository.  Regardless, I'll take a look!

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:47]: Can/should we say something here about which values of `N` we should test?  I understand that we can't be exhaustive here, but I'm thinking it might be a good idea to check:
- When the predicate is true on all work-items
- When the predicate is false on all work-items
- When the predicate deliberately doesn't result in contiguous groups

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:123]: I'm not sure where to document this, but should we only test the cases where `M` >= `N`?

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:120]: Similar to the question above: I think we should make sure we check:
- When the size is 1
- When the size is equal to the sub-group size

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:149]: I'm not sure about `M/N` here (or for `get_group_range()`).  Is `M` here intended as the result of `sub_group::get_local_range()` or `sub_group::get_max_local_range()`?

I think we want the former -- if you agree, a simple fix would be to clarify what we mean by `M` when it's introduced above.

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:47]: I also think you need to add something like:
```suggestion
of the group created with `false` predicate. Let `ballot` be the result of `get_ballot_group`.
```
...or find some way to remove `ballot.` from the descriptions below.

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:123]:
```suggestion
from. Let `fixed` be the result of `get_fixed_size_group`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:169]:
```suggestion
value is equal to `fixed.get_group_id()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:174]:
```suggestion
return value is equal to `fixed.get_local_id()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:179]:
```suggestion
value is equal to `fixed.get_group_range()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:184]:
```suggestion
value is equal to `fixed.get_local_range()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:195]:
```suggestion
The `get_tangle_group` is called with the `sub_group` of the invocation. This
will only be called by `N` items of the sub-group, where `N` is
strictly less than the size of the sub-group.
```

I don't think we can guarantee that it will be the _first_ `N` -- it'll be whichever work-items happen to be active in this control flow.

Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:195]:
```suggestion
The `get_tangle_group` is called with the `sub_group` of the invocation. This
will only be called by the first `N` items of the sub-group, where `N` is
strictly less than the size of the sub-group. Let `tangle` be the result of
`get_tangle_group`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:241]:
```suggestion
value is equal to `tangle.get_group_id()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:246]:
```suggestion
return value is equal to `tangle.get_local_id()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:251]:
```suggestion
value is equal to `tangle.get_group_range()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:256]:
```suggestion
value is equal to `tangle.get_local_range()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:195]: I think it's important for the CTS to deliberately check usage of a tangle in two sides of a branch:
```c++
if (condition) {
  auto tangle = get_tangle_group(sg);
  // checks
}
else {
  auto tangle = get_tangle_group(sg);
  // identical checks
}
```

Compilers are incredibly likely to see code like this and try to be clever by removing the branch, but that ignores the desired `tangle_group` semantics.  If we don't have a test like this, I don't think we can be confident that an implementation actually understands tangles.

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:266]: As above: we should clarify if `M` is `get_local_range()` or `get_max_local_range()`.

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:265]: An implementation could pass this test by implementing `opportunistic_group` as a wrapper around `sub_group`.  I think we need to make sure we deliberately test `get_opportunistic_group` inside of some divergent control flow, similar to tangles.

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:266]:
```suggestion
Let `M` be the size of the sub-group of the invocation. Let `opportunistic` be the result
of `get_opportunistic_group`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:312]:
```suggestion
value is equal to `opportunistic.get_group_id()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:317]:
```suggestion
return value is equal to `opportunistic.get_local_id()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:322]:
```suggestion
value is equal to `opportunistic.get_group_range()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:327]:
```suggestion
value is equal to `opportunistic.get_local_range()` converted to `linear_id_type`.
```

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:344]: This may be a good place to call out some of the specific tests or divergence requirements that I mentioned above, but I'll leave it to you to decide.

**Review by Pennycook — CHANGES_REQUESTED** [review]

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc:149]: I agree, it should be `sub_group::get_local_range()`. I have updated it in all the places it should be clarified.

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc:47]: I think in general we can simply refer to them by the member function names without an object attached, so I have just removed the object from the descriptions.

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc]: First here refers to the way the control-flow is split, i.e. we pick the ones with `sub_group.get_local_linear_id() < N`. I suppose I can clarify that here, but do we need to be that specific in the test plan? I'm fine with either your suggestion or being very specific.

Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc]: Currently my WIP tests use something like:

```
parallel_for(... [&](sycl::nd_item<1> it) {
  if (!condition)
    return;
  auto tangle = get_tangle_group(sg);
  // checks
});
```

Do you think it would be fine to separate them, i.e. something like:

```
parallel_for(... [&](sycl::nd_item<1> it) {
  if (!condition)
    return;
  auto tangle = get_tangle_group(sg);
  // checks
});

...

parallel_for(... [&](sycl::nd_item<1> it) {
  if (condition)
    return;
  auto tangle = get_tangle_group(sg);
  // identical checks
});
```

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc:120]: `size == 1` is fine, but I worry about matching the sub-group size exactly. We are already making some pretty heavy assumptions by picking some `N > 1` where `N` is both a divisor and power-of-two. I suspect `2`, `4` and `8` would be safe picks, but AFAIK nothing stops an implementation from only allowing sub-group size 1. What we can do is test for both `1`, `4`, and `8` and do some checks to make sure they won't hit an unsupported sub-group size.

However, if we were to check with a set `sub_group` size I worry that we have to make a finite list of sub-group sizes, as they have to be specified using `[[sycl::reqd_sub_group_size(...)]]` and in that case we might leave out some hardware that could at least test `1` and probably also `4` and `8`.

Comment (Pennycook) [reply test_plans/non_uniform_groups.asciidoc:120]: It's a shame we don't have named sub-groups yet -- then we could just test the device's preferred sub-group size.

Your suggestion to test `1`, `4` and `8` sounds good to me.

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [reply test_plans/non_uniform_groups.asciidoc]: No, I don't think this is enough.  The behavior we want to test isn't whether tangles can exist for both `condition` and `!condition`, but rather that they can exist _simultaneously_.  If the tests are split out, there's no risk of a compiler optimizing too aggressively.

It might be clearer if I complete my example.  The concern is that a compiler may see this:
```c++
if (condition) {
  auto tangle = get_tangle_group(sg);
  // checks
}
else {
  auto tangle = get_tangle_group(sg);
  // identical checks
}
```

and replace it with this:
```c++
if (condition || !condition) {
  auto tangle = get_tangle_group(sg);
  // checks
}
```

That transformation is legal for most code, but is illegal for code containing tangle groups.

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc:345]: I have added a list with the specific group setups here.

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc:266]: Should be addressed by the list of group-constructions below.

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc]: I am of two minds regarding this. On one hand, I agree that an optimizer might elect to make that optimization, but on the other hand I don't think the CTS should make tests based on potential optimizations. An implementation could simply turn off optimizations in the CTS and pass and there could be a plethora of other patterns that could also be used to cause mayhem here, so it seems a little contrived.

Comment (Pennycook) [reply test_plans/non_uniform_groups.asciidoc]: Oh, I see.  I think this is fine, then -- I'm just not used to reading the test plans.

**Comment by steffenlarsen** [comment]
Draft PR with the tests: https://github.com/KhronosGroup/SYCL-CTS/pull/875

File: test_plans/non_uniform_groups.asciidoc
Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc]: I see what you mean. Would it be enough to do it just for the API testing? I.e. all the algorithm tests can still just use one control flow? I would assume in general having the two paths in one test should be good enough to weed out the worst cases of this kind of optimizations.

Comment (Pennycook) [reply test_plans/non_uniform_groups.asciidoc]: Yes, I think that's sufficient.  That will give us some confidence that the groups are being constructed correctly in weird corner cases.  We're already not going to test all possible code they can put in those branches, so I'm okay if we skip the algorithms too!

Comment (steffenlarsen) [reply test_plans/non_uniform_groups.asciidoc]: Alright! I have tried specifying this in the test plan.

File: test_plans/non_uniform_groups.asciidoc
Comment (Pennycook) [inline test_plans/non_uniform_groups.asciidoc:196]:
```suggestion
some value strictly less than `M`. `get_tangle_group` is called in two split
```

**Review by Pennycook — APPROVED** [review]
Typo aside, LGTM!

File: test_plans/non_uniform_groups.asciidoc
Comment (keryell) [inline test_plans/non_uniform_groups.asciidoc:119]: Are you excluding the case of 1. That seems interesting to test corner cases, otherwise.

Comment (keryell) [reply test_plans/non_uniform_groups.asciidoc:119]: Oh I thought you were running with several N and in that case testing also the boundaries would have made sense.

**Comment by bader** [comment]
@keryell, could you please approve the PR if all your concerns are addressed?

**Review by keryell — APPROVED** [review]
That looks interesting!

**Pull request merged**
